sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py

```diff
@@ -0,0 +1,126 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+import hashlib
+import logging
+from collections import defaultdict
+from typing import Union, List, Dict
+
+from azure.ai.ml.constants._common import AzureMLResourceType
+from azure.ai.ml.entities import Component
+from azure.ai.ml.entities._builders import BaseNode
+
+
+logger = logging.getLogger(__name__)
+
+
+class CachedNodeResolver(object):
+    """Class to resolve component in nodes with cached component resolution results.
+
+    This class is thread-safe if:
+    1) self._resolve_nodes and self._register_node_to_resolve are not called concurrently in the same thread
+    2) self._resolve_component is only called concurrently on independent components
+        a) we have used a component hash to deduplicate components to resolve;
+        b) nodes are registered & resolved layer by layer, so all child components are already resolved
+          when we register a pipeline node;
+        c) potential shared dependencies like compute and data have been resolved before calling self._resolve_nodes.
+    """
+    _ANONYMOUS_HASH_PREFIX = "anonymous_component:"
+    _YAML_SOURCE_PREFIX = "yaml_source:"
+    _ORIGIN_PREFIX = "origin:"
```

Comment:
when is this origin prefix used?

Reply:
no need now as str component has been resolved directly and skipped in node registration.